### ISSUE TITLE
Octahedral field is not smooth enough

### ISSUE THREAD

**Issue by Xiaoyanglib** [Ankbzpx/frame-field#2]
Hello, thank you for your amazing work! I am using the PyTorch version of the code. I tried a simple surface reconstruction on a single tetrahedron, and the reconstruction result is very good. 

However, I observed that the octahedral field is not smooth enough near the feature lines. Is this a bug, or is it because the octahedral field is only an auxiliary field for the signed distance field, the constraints on the octahedral field are not strong enough?

For reference, I traced the `model.octa_mlp` using `torch.jit.trace`; here is the `.pt` file and my dataset.

[tet_octa_traced.zip](https://github.com/user-attachments/files/24234384/tet_octa_traced.zip)

<img width="950" height="1099" alt="Image" src="https://github.com/user-attachments/assets/1f258bb0-e23a-4358-8770-a46c46477bbd" />

<img width="975" height="1104" alt="Image" src="https://github.com/user-attachments/assets/65cde488-ce67-49d2-9071-0808493d5349" />

**Comment by Ankbzpx** [comment]
Good observation!

The main issue is that we write our octahedral variety condition as a soft constraint, such that under stochastic gradient optimization, is unlikely to be satified exactly. In visualization, we still treat the R9 output as if they represent valid octahedral frames, so the results may appear not as smooth.

Here is what I reproduced in JAX using your data (column 2), with octahedral frames evaluated at 128^3 MC vertices.

<img width="1417" height="942" alt="Image" src="https://github.com/user-attachments/assets/364c7f93-5db2-482a-8042-940286cc560b" />

For accurate recovery, we need to project R9 coefficient back to octahedral variety using SDP then perform the power iteration (column 3). We wrapped the [reference SDP implementation](https://github.com/dpa1mer/arff) using [CPP bindings](https://github.com/Ankbzpx/frame-field-utils), that can be called as
```
sh4 = proj_sh4_sdp(sh4)
```
Some remarks:
1. Given we align one representation vector to normal direction and leave other two constrained by smoothness alone, the normal aligned one remains the domain eigenvector and the recovered frames are normal aligned without SDP (as in your figure).
2. Normal aligned octahedral field usually has singularities but appear at tangent plane, so the normal alignment is intacted and the regularization is still effective.
3. We use pointwise analytical gradient norm to optimize smoothness, and given the high capacity of SIREN, some tangential twists can be push to a very small region. We are not aware how to control this behavior.

**Comment by Xiaoyanglib** [comment]
Thank you for your explanation! 

The field I showed earlier has already been projected (using this [code](https://github.com/cgg-bern/AlgoHex/blob/9fe0fb9c9732947ac4304cba86c458c04dd8da4a/src/AlgoHex/SphericalHarmonics/SHProjectorRay.hh#L41)).  

Meanwhile, I observed that in the result you presented (column 3), singularities appear near the feature edges, whereas some traditional methods (e.g., [Practical 3D frame field generation](https://dl.acm.org/doi/10.1145/2980179.2982408)) tend to produce singularities farther away from the feature edges. What is the reason for this?

<img width="2042" height="918" alt="Image" src="https://github.com/user-attachments/assets/83087fb3-a489-4e01-b4a6-a67609fc8154" />

**Comment by Ankbzpx** [comment]
Nice library! I think you should use [SHProjectorSDP](https://github.com/cgg-bern/AlgoHex/blob/9fe0fb9c9732947ac4304cba86c458c04dd8da4a/src/AlgoHex/SphericalHarmonics/SHProjectorSDP.hh#L28) before applying Ray's method. (In my case, I don't have access to the MOSEK solver, so I wrap mine with SCS instead.)

In terms of frame field difference, I believe it is mainly due to [sample weight](https://github.com/Ankbzpx/frame-field/blob/00b8a67e43b68bc114c5b7848e603a2b275dc925/pytorch/train_pytorch.py#L130). In our application, we want to smooth along the zero level set, but we cannot directly do that, so we use a soft relaxation. So essentially, we are closer to "Octahedral Frames for Feature-Aligned Cross-Fields" by Zhang et al, which uses a manifold Laplacian as opposed to the Volume one from Ray et al.

IMO, for volume frame field design, you could try to remove that sample weight, so even replace smoothness loss with LipMLP instead.

**Comment by Xiaoyanglib** [comment]
Thanks for your help and detailed explanation!
